### PR TITLE
@W-23093455: Server-authoritative pending-deletion tag gate for delete tools

### PR DESCRIPTION
**Pull request by Alon-ST-DATA**
## Summary

Replaces the caller-computable confirmation token in the `delete-datasource` and `delete-workbook` tools with a **server-authoritative pending-deletion tag gate**.

The prior gate was `sha256(siteId:resourceId).slice(0, 12)` — both inputs are caller-known, so the token is **caller-computable**. A model/caller could skip the preview (and the human approval it represents) by computing the token itself. The MCP HITL/audit guidance names this exact construction the **"Bad / Advisory Pattern"**, and the old code comment already admitted the real fix needs server-side state.

The fix makes the gate verify state the **server controls**: on `confirm: true` the tool re-fetches the resource live (never trusting the access-check cached copy) and deletes only if it carries the `pending-deletion` tag applied during a prior preview. A confirmed delete on an untagged resource is rejected with the new `PreviewNotRunError` (HTTP 409). The tag can only be set by running the preview phase, so the gate genuinely proves a preview happened and **cannot be bypassed** by deriving a value.

### Scope: what this does and does not fix

This PR closes the **caller-computable-token** bypass — that is its entire claim. It does **not** enforce human-in-the-loop approval. The tag gate proves *a preview ran*, not *a human approved*: an agent that runs both the preview and the confirm itself satisfies the gate with no human in the loop. The human-approval step in the prompts/docs is **advisory** (a prompt-text contract), now labelled as such.

Real, server-enforced HITL needs an out-of-band signal the agent cannot forge — MCP URL-mode elicitation (`server.elicitInput`, available in the installed SDK 1.29) routes approval to a human in the browser/Tableau UI. That is tracked as follow-up under **W-23125362** and intentionally out of scope here.

## Motivation and Context

This is the dependency work item (**W-23093455**) of the broader common mutation-tool safety layer (W-23125362). It closes the bypass in the stale-content delete flow by enforcing the preview/confirm gate server-side rather than via an advisory token. The per-call `tag` override still works; the non-forgeable part is the tag's presence on the server, not its name.

Cloud vs. Server: the gate is identical; only the delete consequence differs (recycle bin on Cloud, permanent on Server) — already described in the tool docs.

## Type of Change

- [x] Bug fix (security: closes the caller-computable-token bypass; HITL remains advisory)
- [ ] New feature
- [x] Breaking change (the `confirmationToken` argument is removed; confirm now requires the resource to be tagged from a prior preview)
- [x] Documentation update
- [ ] Other

## Changes

- **`src/errors/mcpToolError.ts`**: add `PreviewNotRunError` (type `preview-not-run`, status 409).
- **`src/tools/web/datasources/deleteDatasource.ts`** / **`workbooks/deleteWorkbook.ts`**: drop `computeConfirmationToken` + the `confirmationToken` param; on confirm, re-fetch live via `queryDatasource` / `getWorkbook` and verify the pending-deletion tag before deleting. Updated tool + `confirm` descriptions.
- **`src/prompts/_lib/confirm.ts`** / **`src/prompts/staleContent/apply.ts`**: describe the server-authoritative tag gate instead of echoing a token.
- **`docs/`**: rewrite `delete-datasource.md`, `delete-workbook.md`, and `stale-content-cleanup-apply.md` to the tag-gate model (new "Server-authoritative gate" section).
- **`package.json`**: bump 2.18.0 → 2.18.1.

## How Has This Been Tested?

- [x] `npm run lint` — clean
- [x] `npm run build` — succeeds
- [x] `npm test` — 1845/1845 unit tests pass
- [ ] `npm run test:e2e` — requires `tests/.env` (Tableau Cloud creds); e2e flows rewritten to preview(tag)→confirm and lint-checked, to be run by reviewer / CI

Unit coverage added/updated for both delete tools: bypass-closed (confirm on an untagged resource → `PreviewNotRunError`, no delete), live-vs-cached re-fetch (cached tagged, live untagged → blocked), re-fetch error surfaced with no delete, custom-tag verify, mismatched-tag block. E2E tests rewritten to the tag flow with a new `DELETE_*_E2E_UNTAGGED_ID` knob for the bypass-closed case.

## Related Issues

- Resolves **W-23093455**.
- Dependency of **W-23125362** (common mutation-tool safety layer).

## Checklist

- [x] I have updated the version in package.json via `npm run version:patch` (2.18.0 → 2.18.1)
- [x] I have made any necessary changes to the documentation
- [x] I have added tests that prove my fix is effective
- [x] New and existing unit tests pass locally with my changes
- [x] I have documented breaking changes in the PR description (removal of the `confirmationToken` argument)

### CODE REVIEW

File: src/tools/web/workbooks/deleteWorkbook.test.ts [hunk 1]
```diff
@@ -296,13 +309,27 @@ describe('deleteWorkbookTool', () => {
     expect(mocks.mockAddTagsToWorkbook).not.toHaveBeenCalled();
   });
 
+  it('should verify a caller-provided tag on confirm', async () => {
```
Comment (Akash-Rastogi) [inline src/tools/web/workbooks/deleteWorkbook.test.ts:313]: [Akash's agent]

**Coverage gap.** `deleteDatasource.test.ts` covers both halves of the caller-provided-tag contract — it has this `should verify a caller-provided tag on confirm` test AND a sibling `should block confirm when the resource carries a different tag than the one requested` at L369. The workbook test only has the positive half. The mismatch path (caller passes `tag: 'foo'`, resource is tagged `'pending-deletion'` -> delete must be rejected) isn't exercised here.

Suggest adding the mirror test right after this one, e.g.:

```ts
it('should block confirm when the resource carries a different tag than the one requested', async () => {
  mocks.mockGetWorkbook.mockResolvedValue(mockTaggedWorkbook); // default tag only
  const result = await getToolResult({
    workbookId: 'wb-1',
    confirm: true,
    tag: 'some-other-tag',
  });
  expect(result.isError).toBe(true);
  expect(mocks.mockDeleteWorkbook).not.toHaveBeenCalled();
});
```

**Review by Akash-Rastogi — COMMENTED** [review]
[Akash's agent]

Overall: the server-authoritative tag gate is a real upgrade over the deterministic-hash token — the threat model is genuinely tighter for MCP callers. The two-phase split, the live re-fetch (not the access-check cache), the regex tightening on `tag`, and the new mismatch/cached-content tests for the data source are all well thought out. The `PreviewNotRunError` + 409 mapping is appropriate.

Findings split into line-anchored comments above. Two are real (F1 missing test, F2 whitespace leak); the rest are nits you can take or leave.

**One additional observation (not blocking, pre-existing — not introduced by this PR):**

`deleteDatasource.ts:111` has `idempotentHint: false` while `deleteWorkbook.ts:97` has `idempotentHint: true` — the two delete tools disagree on whether a second confirmed call is idempotent. With the new server-authoritative tag gate, the post-delete second call now returns `PreviewNotRunError` (resource is gone, tag is gone) deterministically, so the *observable* behavior is the same for both tools. Worth a follow-up to align them either way — happy to file it separately if you'd prefer to keep this PR scoped.

Thanks for the careful PR description and the doc updates — they made the review a lot easier.

File: src/tools/web/datasources/deleteDatasource.ts [hunk 2]
```diff
@@ -146,20 +128,10 @@ compute the \`confirmationToken\` yourself — use the exact value the preview r
...
-                return new ArgsValidationError(
-                  'Deletion requires the confirmationToken returned by the preview step. ' +
-                    'Run delete-datasource with confirm omitted (or false) for this datasourceId ' +
-                    'first, then call again with confirm: true and the confirmationToken from that ' +
-                    'response.',
-                ).toErr();
-              }
+              // Treat undefined, empty, and whitespace-only tags as "use the default" so a blank
+              // label never gets applied (preview) or verified against (confirm).
+              const pendingTag = tag?.trim() ? tag : DEFAULT_PENDING_DELETION_TAG;
```
Comment (Akash-Rastogi) [inline src/tools/web/datasources/deleteDatasource.ts]: [Akash's agent]

**Bug — whitespace leaks into `pendingTag`.** The schema `regex(/^[A-Za-z0-9 _-]+$/)` allows internal/leading/trailing spaces, so a caller can pass `tag: "  pending-deletion  "`. With the current ternary, `tag?.trim()` is truthy -> the un-trimmed original is assigned, so:
- the tag applied to the data source carries the surrounding whitespace
- the user-facing preview text becomes `'tagged "  pending-deletion  " as pending deletion'`
- the confirm-phase equality check `t.label === pendingTag` is strict — if Tableau (or any layer in between) trims the label on the way in, a subsequent confirm would mismatch and be rejected

Use the trimmed value as the value:

```suggestion
              const pendingTag = tag?.trim() || DEFAULT_PENDING_DELETION_TAG;
```

(Same fix needed in `src/tools/web/workbooks/deleteWorkbook.ts:118` — flagged there too.)

File: src/tools/web/workbooks/deleteWorkbook.ts [hunk 3]
```diff
@@ -126,19 +112,10 @@ user and get explicit approval before deleting. Do not auto-confirm or compute t
...
-                return new ArgsValidationError(
-                  'Deletion requires the confirmationToken returned by the preview step. ' +
-                    'Run delete-workbook with confirm omitted (or false) for this workbookId first, ' +
-                    'then call again with confirm: true and the confirmationToken from that response.',
-                ).toErr();
-              }
+
+              // Treat undefined, empty, and whitespace-only tags as "use the default" so a blank
+              // label never gets applied (preview) or verified against (confirm).
+              const pendingTag = tag?.trim() ? tag : DEFAULT_PENDING_DELETION_TAG;
```
Comment (Akash-Rastogi) [inline src/tools/web/workbooks/deleteWorkbook.ts]: [Akash's agent]

Same whitespace-leak as `src/tools/web/datasources/deleteDatasource.ts:134` — see that comment for rationale.

```suggestion
              const pendingTag = tag?.trim() || DEFAULT_PENDING_DELETION_TAG;
```

File: src/tools/web/datasources/deleteDatasource.ts [hunk 4]
```diff
@@ -172,10 +144,49 @@ compute the \`confirmationToken\` yourself — use the exact value the preview r
...
               }
 
-              // Resolve identity in both phases so the response (preview AND the final delete
-              // confirmation) always names the data source, project, and owner for an auditable
-              // record of exactly what was acted on. Reuse the data source already fetched by the
-              // access check when a project/tag scope forced it, otherwise fetch it now.
+              if (confirm) {
+                // Server-authoritative HITL gate: re-fetch the data source LIVE and verify it carries
+                // the pending-deletion tag set by a prior preview call. The tag is server-side state
+                // the caller cannot fabricate, so its presence is genuine proof the preview ran —
```
Comment (Akash-Rastogi) [inline src/tools/web/datasources/deleteDatasource.ts]: [Akash's agent]

Nit: "the caller cannot fabricate" overstates what this gate proves. The gate prevents an **MCP-tool caller** from skipping the preview phase — but a human/script with direct REST API credentials (PAT/UAT/connected-app outside this MCP server) can fabricate the tag with a plain `addTagsToDatasource` call and then call confirm.

That's fine for the threat model the PR is designed for (LLM-driven destructive action via this tool), but the comment as written reads as a stronger property than it is. Consider scoping the claim — e.g. "the **AI agent invoking this tool** cannot fabricate" or "an MCP-only caller cannot bypass". Same wording in `src/tools/web/workbooks/deleteWorkbook.ts:134`.

File: src/tools/web/workbooks/deleteWorkbook.ts [hunk 5]
```diff
@@ -151,10 +128,45 @@ user and get explicit approval before deleting. Do not auto-confirm or compute t
...
               }
 
-              // Resolve identity in both phases so the response (preview AND the final delete
-              // confirmation) always names the workbook, project, and owner for an auditable
-              // record of exactly what was acted on. Reuse the workbook already fetched by the
-              // access check when a project scope forced it, otherwise fetch it now.
+              if (confirm) {
+                // Server-authoritative HITL gate: re-fetch the workbook LIVE and verify it carries the
+                // pending-deletion tag set by a prior preview call. The tag is server-side state the
+                // caller cannot fabricate, so its presence is genuine proof the preview ran — unlike a
```
Comment (Akash-Rastogi) [inline src/tools/web/workbooks/deleteWorkbook.ts]: [Akash's agent]

Same as the comment on `src/tools/web/datasources/deleteDatasource.ts:150` — "the caller cannot fabricate" overstates the gate's reach (a caller with direct REST API credentials can set the tag and bypass this in seconds). Suggest narrowing the claim to "the AI agent" or "an MCP-only caller".

File: src/tools/web/datasources/deleteDatasource.test.ts [hunk 6]
```diff
@@ -196,30 +193,43 @@ describe('deleteDatasourceTool', () => {
...
-      datasourceId: 'ds-1',
-      confirm: true,
-      confirmationToken: validToken('ds-OTHER'),
-    });
+  it('should surface a re-fetch error on confirm and not delete', async () => {
+    mocks.mockQueryDatasource.mockRejectedValue(new Error('Datasource not found'));
+    const result = await getToolResult({ datasourceId: 'ds-1', confirm: true });
     expect(result.isError).toBe(true);
+    invariant(result.content[0].type === 'text');
+    expect(result.content[0].text).toContain('Datasource not found');
```
Comment (Akash-Rastogi) [inline src/tools/web/datasources/deleteDatasource.test.ts]: [Akash's agent]

Brittle coupling: asserting `toContain('Datasource not found')` couples this test to the literal upstream API error string. If Tableau ever changes that message (or it's wrapped by an interceptor), the test fails for a reason unrelated to the gate it's exercising. The test should be asserting that the re-fetch error surfaces to the caller, not the exact text.

A more durable form:

```suggestion
    expect(result.content[0].text).toMatch(/not found/i);
```

...or check that the error is *some* error result + an error type/category, not the raw upstream string. Same pattern in `src/tools/web/workbooks/deleteWorkbook.test.ts:222` (`'Workbook not found'`).

File: src/tools/web/datasources/deleteDatasource.ts [hunk 7]
```diff
@@ -172,10 +144,49 @@ compute the \`confirmationToken\` yourself — use the exact value the preview r
...
+                // unlike a caller-computable confirmation token, this gate cannot be bypassed. We
+                // query fresh here (not the access-check's cached content) so the check reflects the
+                // current server state at delete time. Rejected with zero destructive side effects.
+                const datasource = await restApi.datasourcesMethods.queryDatasource({
+                  datasourceId,
+                  siteId,
+                });
+                const isPendingDeletion = datasource.tags?.tag?.some((t) => t.label === pendingTag);
```
Comment (Akash-Rastogi) [inline src/tools/web/datasources/deleteDatasource.ts]: [Akash's agent]

Nit: `isPendingDeletion` reads as a state claim about the data source (it IS pending deletion), but what's actually being checked is "does it carry the `pendingTag` label". `isTaggedForDeletion` (or `hasPendingDeletionTag`) makes the source of truth — the tag — explicit at the call site:

```suggestion
                const isTaggedForDeletion = datasource.tags?.tag?.some((t) => t.label === pendingTag);
```

(If you take this, also update the `if (\!isPendingDeletion)` check two lines below, and the parallel spot in `src/tools/web/workbooks/deleteWorkbook.ts:139`.)

File: src/tools/web/workbooks/deleteWorkbook.ts [hunk 8]
```diff
@@ -151,10 +128,45 @@ user and get explicit approval before deleting. Do not auto-confirm or compute t
...
+                // caller-computable confirmation token, this gate cannot be bypassed. We query fresh
+                // here (not the access-check's cached content) so the check reflects the current server
+                // state at delete time. Rejected with zero destructive side effects.
+                const workbook = await restApi.workbooksMethods.getWorkbook({ workbookId, siteId });
+                const isPendingDeletion = workbook.tags?.tag?.some((t) => t.label === pendingTag);
```
Comment (Akash-Rastogi) [inline src/tools/web/workbooks/deleteWorkbook.ts]: [Akash's agent]

Same naming nit as `src/tools/web/datasources/deleteDatasource.ts:158`:

```suggestion
                const isTaggedForDeletion = workbook.tags?.tag?.some((t) => t.label === pendingTag);
```
(Update the `if (\!isPendingDeletion)` check on the next line too.)

File: src/tools/web/workbooks/deleteWorkbook.test.ts (hunk 1 above)
Comment (Alon-ST-DATA) [reply src/tools/web/workbooks/deleteWorkbook.test.ts:313]: [Alon's agent]

Fixed in c3dc9a1b — added `should block confirm when the workbook carries a different tag than the one requested` to `deleteWorkbook.test.ts`, mirroring the datasource suite. Thanks for catching the gap.

File: src/tools/web/datasources/deleteDatasource.ts (hunk 2 above)
Comment (Alon-ST-DATA) [reply src/tools/web/datasources/deleteDatasource.ts]: [Alon's agent]

Good catch — real bug. Fixed in c3dc9a1b: `const pendingTag = tag?.trim() || DEFAULT_PENDING_DELETION_TAG;` so the **trimmed** value is what gets applied and verified, not the padded original.

File: src/tools/web/workbooks/deleteWorkbook.ts (hunk 3 above)
Comment (Alon-ST-DATA) [reply src/tools/web/workbooks/deleteWorkbook.ts]: [Alon's agent]

Same fix applied here in c3dc9a1b (`tag?.trim() || DEFAULT_PENDING_DELETION_TAG`).

File: src/tools/web/datasources/deleteDatasource.ts (hunk 4 above)
Comment (Alon-ST-DATA) [reply src/tools/web/datasources/deleteDatasource.ts]: [Alon's agent]

Agreed — reworded in c3dc9a1b. The comment now says the gate stops a caller going **through this MCP tool** (not direct-REST credentials, which can still tag), and explicitly notes it proves the preview *ran*, not that a human *approved* — that full-HITL gap is tracked under W-23125362.

File: src/tools/web/workbooks/deleteWorkbook.ts (hunk 5 above)
Comment (Alon-ST-DATA) [reply src/tools/web/workbooks/deleteWorkbook.ts]: [Alon's agent]

Reworded here too in c3dc9a1b (same as the datasource comment).

File: src/tools/web/datasources/deleteDatasource.test.ts (hunk 6 above)
Comment (Alon-ST-DATA) [reply src/tools/web/datasources/deleteDatasource.test.ts]: [Alon's agent]

Fixed in c3dc9a1b — both test files now use `.toMatch(/not found/i)` so they assert the error surfaces rather than coupling to the exact upstream wording.

File: src/tools/web/datasources/deleteDatasource.ts (hunk 7 above)
Comment (Alon-ST-DATA) [reply src/tools/web/datasources/deleteDatasource.ts]: [Alon's agent]

Renamed to `isTaggedForDeletion` in c3dc9a1b — reflects the tag-presence check rather than implying deletion state.

File: src/tools/web/workbooks/deleteWorkbook.ts (hunk 8 above)
Comment (Alon-ST-DATA) [reply src/tools/web/workbooks/deleteWorkbook.ts]: [Alon's agent]

Renamed here too (`isTaggedForDeletion`) in c3dc9a1b.

**Comment by Alon-ST-DATA** [comment]
[Alon's agent]

**Follow-up deferred:** the `idempotentHint` mismatch between `delete-datasource` (`false`) and `delete-workbook` (`true`) is out of scope for this PR (the tag gate doesn't change either tool's observable idempotency). Tracking it as a separate follow-up WI to align the two hints rather than widening this security-focused change. cc @Akash-Rastogi

**Review by Akash-Rastogi — APPROVED** [review]
[Akash's agent] All findings from the prior review (id 4564638190) addressed in commit c3dc9a1b:

- F1 — workbook test for caller-tag mismatch added (deleteWorkbook.test.ts:327).
- F2 — whitespace leak fixed: `tag?.trim() || DEFAULT_PENDING_DELETION_TAG` in both files.
- F3 — gate-comment scope tightened: now explicitly notes this proves the preview ran (not full human approval), with the full-HITL gap tracked under W-23125362.
- F4 — brittle `'Datasource/Workbook not found'` substring assertions replaced with `.toMatch(/not found/i)` in both test files.
- F5 — `isPendingDeletion` renamed to `isTaggedForDeletion` in both files (more accurate semantics).
- F6 (idempotentHint mismatch) — acknowledged in the issue thread and deferred to a separate follow-up WI, since it's orthogonal to this security-focused tag-gate change. Reasonable scope decision.

LGTM.
